File: Java/bCAPClient/Src/orin2/bcap/BCAPConnectionTCP.java
/**
 * Software License Agreement (MIT License)
 *
 * @copyright Copyright (c) 2015 DENSO WAVE INCORPORATED
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package orin2.bcap;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.net.Socket;
import java.net.InetSocketAddress;

import orin2.library.HResult;
import orin2.library.ORiN2Exception;

class BCAPConnectionTCP extends BCAPConnectionBase implements Runnable {
	private static final int PROC_INIT       = 1;
	private static final int PROC_RELEASE    = 2;
	private static final int PROC_SENDRECV   = 3;
	private static final int PROC_SETTIMEOUT = 4;
	
	private Socket              m_sock   = null;
	private DataInputStream     m_in     = null;
	private DataOutputStream    m_out    = null;
	private BCAPPacketConverter m_conv   = new BCAPPacketConverter();
	private Object[]            m_args   = null;
	private BCAPPacket          m_packet = null;
	private Throwable           m_except = null;
	
	public BCAPConnectionTCP(String host, int port, int timeout, int retry) throws Throwable {
		m_args = new Object[]{PROC_INIT, host, port, timeout, retry};
		
		Thread th = new Thread(this);		
		th.start();
		WaitThread(th);
				
		m_args = null;
		if(m_except != null) throw m_except;
		
		SetTimeout(timeout);
		SetRetry(retry);
	}
	
	@Override
	public void Release() {
		m_args = new Object[]{PROC_RELEASE};
		
		Thread th = new Thread(this);
		th.start();
		WaitThread(th);
		
		m_args = null;
	}
	
	@Override
	public synchronized BCAPPacket SendRecv(BCAPPacket packet) throws Throwable {
		m_args = new Object[]{PROC_SENDRECV, packet};
		
		Thread th = new Thread(this);
		th.start();
		WaitThread(th);
		
		m_args = null;
		if(m_except != null) throw m_except;
		return m_packet;
	}
	
	@Override
	public void SetTimeout(int timeout) throws Throwable {
		m_args = new Object[]{PROC_SETTIMEOUT, timeout};
		
		Thread th = new Thread(this);		
		th.start();
		WaitThread(th);
		
		m_args = null;
		if(m_except != null) throw m_except;
	}
	
	@Override
	public void SetRetry(int retry) throws Throwable {
		super.SetRetry(1);
	}
	
	@Override
	public void run() {
		m_except = null;
		try {
			switch((Integer)m_args[0]){
				case PROC_INIT:
					th_init((String)m_args[1], (Integer)m_args[2], (Integer)m_args[3]);
					break;
				case PROC_RELEASE:
					th_release();
					break;
				case PROC_SENDRECV:
					th_sendrecv((BCAPPacket)m_args[1]);
					break;
				case PROC_SETTIMEOUT:
					th_puttimeout((Integer)m_args[1]);
					break;
			}
		} catch(Throwable cause) {
			m_except = cause;
		}
	}
	
	private void th_init(String host, int port, int timeout) throws Throwable {
		m_sock = new Socket();
		m_sock.connect(new InetSocketAddress(host, port), timeout);
		m_sock.setTcpNoDelay(true);
		m_in = new DataInputStream(m_sock.getInputStream());
		m_out = new DataOutputStream(m_sock.getOutputStream());
	}

	private void th_release(){
		try { m_out.close(); }
		catch(Throwable ignore) { ; }
		finally { m_out = null; }
		
		try { m_in.close(); }
		catch(Throwable ignore) { ; }
		finally { m_in = null; }
		
		try { m_sock.close(); }
		catch(Throwable ignore) { ;	}
		finally { m_sock = null; }
	}
	
	private void th_sendrecv(BCAPPacket packet) throws Throwable {
		byte[] bRecv, bLen = new byte[4];
		m_packet = new BCAPPacket();
		
		// Send packet
		packet.Serial = m_serial;
		packet.Reserv = 0;
		m_out.write(m_conv.Serialize(packet));
		
		// Increments serial number
		m_serial = (short)((m_serial != -1) ? m_serial + 1 : 1);
		
		while(true) {
			int soh = m_in.read(); // SOH
			
			m_in.read(bLen); // Length
			int len = (Integer)BCAPByteConverter.Byte2Object(bLen, Integer.class);
			
			bRecv = new byte[len];
			m_in.readFully(bRecv, 5, len - 5);
			
			// Copy SOH and Length
			bRecv[0] = (byte)soh;
			System.arraycopy(bLen, 0, bRecv, 1, 4);
			
			m_packet = m_conv.Deserialize(bRecv);
			if((packet.Serial == m_packet.Serial)
					&& (m_packet.FuncID != HResult.S_EXECUTING)){
				break;
			}
		}
		
		if(HResult.FAILED(m_packet.FuncID)) {
			throw new ORiN2Exception(m_packet.FuncID);
		}
	}
	
	private void th_puttimeout(int timeout) throws Throwable {
		if(timeout < 0)
		{
			m_sock.setSoTimeout(0);
			super.SetTimeout(0);
		}
		else
		{
			m_sock.setSoTimeout(timeout);
			super.SetTimeout(timeout);
		}
	}
}

File: CSharp/bCAPClient/Src/bCAPConnectionTCP.cs
/**
 * Software License Agreement (MIT License)
 *
 * @copyright Copyright (c) 2015 DENSO WAVE INCORPORATED
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using ORiN2.Library;

namespace ORiN2.bCAP
{
    class bCAPConnectionTCP : bCAPConnectionBase
    {
        private TcpClient m_tcpStream;

        private bCAPPacketConverter m_conv;
        private object m_lock;

        public bCAPConnectionTCP()
        {

        }

        public override void Connect(ConnOptEther optEth, int iTimeout, int iRetry)
        {
            try
            {
                m_tcpStream = new TcpClient();
                var iar = m_tcpStream.BeginConnect(optEth.DstAddr, optEth.DstPort, null, null);
                if (!iar.IsCompleted)
                {
                    if (!iar.AsyncWaitHandle.WaitOne(iTimeout))
                    {
                        try
                        {
                            m_tcpStream.Close();
                            m_tcpStream.EndConnect(iar);
                        }
                        catch (Exception) { }
                        throw new TimeoutException();
                    }
                }
                m_tcpStream.EndConnect(iar);

                m_tcpStream.NoDelay = true;
                SetTimeout(iTimeout);
                SetRetry(iRetry);

                m_conv = new bCAPPacketConverter();
                m_lock = new object();
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public override void Dispose()
        {
            if (m_tcpStream.Connected)
            {
                m_tcpStream.GetStream().Dispose();
                m_tcpStream.Close();
            }
            base.Dispose();
        }

        public override void SetTimeout(int iTimeout)
        {
            m_tcpStream.SendTimeout = iTimeout;
            m_tcpStream.ReceiveTimeout = iTimeout;
            base.SetTimeout(iTimeout);
        }

        public override void SetRetry(int iRetry)
        {
            base.SetRetry(1);
        }

        public override void SendPacket(bCAPPacket send)
        {
            byte[] bySend = m_conv.Serialize(send);
            m_tcpStream.GetStream().Write(bySend, 0, bySend.Length);
        }

        public override bCAPPacket RecvPacket(UInt16 iSerial)
        {
            byte[] byRecv;
            var recv = new bCAPPacket();

            using (var ms = new MemoryStream())
            {
                var br = new BinaryReader(m_tcpStream.GetStream());

                while (true)
                {
                    try
                    {
                        // ストリームを初期化
                        ms.Position = 0;
                        ms.SetLength(0);

                        // SOH
                        ms.WriteByte(br.ReadByte());

                        // Message Length
                        byRecv = br.ReadBytes(4);
                        ms.Write(byRecv, 0, byRecv.Length);

                        // 残りを受信
                        byRecv = br.ReadBytes((Int32)bCAPByteConverter.Byte2Value(byRecv, typeof(Int32)) - 5);
                        ms.Write(byRecv, 0, byRecv.Length);

                        // 受信パケットが送信パケットの応答であり、かつ、S_EXECUTINGでなければ終わり
                        recv = m_conv.Deserialize(ms.ToArray());
                        if ((recv.iSerial == iSerial) && (recv.iFuncID != HResult.S_EXECUTING))
                        {
                            break;
                        }
                    }
                    catch (Exception ex)
                    {
                        throw ex;
                    }
                }
            }

            return recv;
        }

        public override bCAPPacket SendAndRecv(bCAPPacket send)
        {
            var recv = new bCAPPacket();

            lock (m_lock)
            {
                // シリアル番号を設定
                send.iSerial = m_iSerial;
                send.iReserv = 0;

                // パケット送信
                SendPacket(send);

                // シリアル番号を更新
                m_iSerial = (UInt16)((m_iSerial != UInt16.MaxValue) ? (m_iSerial + 1) : 1);

                // パケットを受信
                recv = RecvPacket(send.iSerial);
            }

            if (recv.iFuncID < 0)
            {
                throw new ORiN2Exception(recv.iFuncID, null);
            }

            return recv;
        }
    }
}
